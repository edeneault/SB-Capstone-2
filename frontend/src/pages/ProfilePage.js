import React, { useState, useEffect } from "react";
import { LinkContainer } from "react-router-bootstrap";
import { Container, Form, Button, Row, Col, Table } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import Loader from "../components/Loader";
import { getUserDetails, updateUserProfile } from "../actions/userActions";
import { listMyOrders } from "../actions/orderActions";
import { USER_UPDATE_PROFILE_RESET } from "../constants/userConstants";

import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";

toast.configure();

const ProfilePage = ({ location, history }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [name, setName] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");
  const [message, setMessage] = useState(null);

  const dispatch = useDispatch();

  const userDetails = useSelector((state) => state.userDetails);
  const { loading, error, user } = userDetails;

  const userLogin = useSelector((state) => state.userLogin);
  const { userInfo } = userLogin;

  const userUpdateProfile = useSelector((state) => state.userUpdateProfile);
  const { success } = userUpdateProfile;

  const orderListMy = useSelector((state) => state.ordersMyList);
  const { loading: loadingOrders, error: errorOrders, orders } = orderListMy;

  useEffect(() => {
    if (!userInfo) {
      history.push("/login");
    } else {
      if (!user || !user.name || success) {
        dispatch({ type: USER_UPDATE_PROFILE_RESET });
        dispatch(getUserDetails("profile"));
        dispatch(listMyOrders());
      } else {
        setName(user.name);
        setEmail(user.email);
      }
    }
  }, [dispatch, userInfo, history, user, success]);

  const submitHandler = (e) => {
    e.preventDefault();
    if (password !== confirmPassword) {
      setMessage("Passwords do not match");
    } else {
      dispatch(updateUserProfile({ id: user._id, name, email, password }));
    }
  };

  return (
    <Container fluid className='px-5 pt-3 fade-in'>
      <Row>
        <Col md={3}>
          <h2 className=' pb-0 mb-0 slide-in'>
            User Profile<p>Update</p>
          </h2>

          {message && toast(`${error}`, { type: "error" })}
          {error && toast(`${error}`, { type: "error" })}
          {success && toast(`Profile Updated`, { type: "success" })}
          {loading && <Loader />}
          <Form onSubmit={submitHandler}>
            <Form.Group controlId='name'>
              <Form.Label>Email Name</Form.Label>
              <Form.Control
                type='name'
                placeholder='Enter name'
                value={name}
                onChange={(e) => setName(e.target.value)}
              ></Form.Control>
            </Form.Group>
            <Form.Group controlId='email'>
              <Form.Label>Email Address</Form.Label>
              <Form.Control
                type='email'
                placeholder='Enter email'
                value={email}
                onChange={(e) => setEmail(e.target.value)}
              ></Form.Control>
            </Form.Group>

            <Form.Group controlId='password'>
              <Form.Label>Password</Form.Label>
              <Form.Control
                type='password'
                placeholder='Enter password'
                value={password}
                onChange={(e) => setPassword(e.target.value)}
              ></Form.Control>
            </Form.Group>

            <Form.Group controlId='confirmPassword'>
              <Form.Label>Confirm Password</Form.Label>
              <Form.Control
                type='confirmPassword'
                placeholder='Confirm password'
                value={confirmPassword}
                onChange={(e) => setConfirmPassword(e.target.value)}
              ></Form.Control>
            </Form.Group>

            <Button type='submit' variant='primary rounded my-2'>
              Update
            </Button>
          </Form>
        </Col>

        <Col md={9}>
          <h2>My Orders</h2>
          {loadingOrders ? (
            <Loader />
          ) : errorOrders ? (
            toast(`${errorOrders}`, { type: "error" })
          ) : (
            <Table bordered hover responsive className='table-sm'>
              <thead>
                <tr className='bg-black text-light'>
                  <th>ID</th>
                  <th>DATE</th>
                  <th>TOTAL</th>
                  <th>PAID</th>
                  <th>DELIVERED</th>
                  <th>DETAILS</th>
                </tr>
              </thead>
              <tbody>
                {orders.map((order) => (
                  <tr key={order._id}>
                    <td>{order._id}</td>
                    <td>{order.createdAt.substring(0, 10)}</td>
                    <td>{order.totalPrice}</td>
                    <td>
                      {order && order.isPaid ? (
                        order.paidAt.substring(0, 10)
                      ) : (
                        <i
                          className='fas fa-times'
                          style={{ color: "red" }}
                        ></i>
                      )}
                    </td>
                    <td>
                      {order && order.isDelivered ? (
                        order.deliveredAt.substring(0, 10)
                      ) : (
                        <i
                          className='fas fa-times'
                          style={{ color: "red" }}
                        ></i>
                      )}
                    </td>
                    <td>
                      <LinkContainer to={`/orders/${order._id}`}>
                        <Button variant='light' className='btn-sm'>
                          Details
                        </Button>
                      </LinkContainer>
                    </td>
                  </tr>
                ))}
              </tbody>
            </Table>
          )}
        </Col>
      </Row>
    </Container>
  );
};

export default ProfilePage;
