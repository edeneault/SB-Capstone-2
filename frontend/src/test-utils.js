import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

// create a customRender that wraps the UI in a memory Router

const MemoryRouterWithInitialRoutes = ({ children, initialRoutes }) => {
  return <MemoryRouter initialEntries={initialRoutes}>{children}</MemoryRouter>;
};

const customRender = (ui, options) => {
  const initialRoutes =
    options && options.initialRoutes ? options.initialRoutes : ["/"];

  return render(ui, {
    wrapper: (args) =>
      MemoryRouterWithInitialRoutes({
        ...args,
        initialRoutes,
      }),
    ...options,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
