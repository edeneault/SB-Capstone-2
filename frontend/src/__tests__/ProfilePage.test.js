import React from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { initialState, render } from "../test-utils";
import ProfilePage from "../pages/ProfilePage.js";
import { createMemoryHistory } from "history";

const mockStore = configureMockStore([thunk]);

describe("test", () => {
  const history = createMemoryHistory();
  const path = `/profile`;

  test("renders ProfilePage without crashing", () => {
    const store = mockStore(initialState);

    const { getByText } = render(
      <Provider store={store}>
        <ProfilePage history={history} />
      </Provider>,
      { initialRoutes: [`/profile`] },
    );

    expect(getByText(/My Orders/i)).toBeInTheDocument();
  });

  test("matches snapshot", function () {
    const store = mockStore(initialState);

    const { asFragment } = render(
      <Provider store={store}>
        <ProfilePage history={history} />
      </Provider>,
      { initialRoutes: [`/profile`] },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
