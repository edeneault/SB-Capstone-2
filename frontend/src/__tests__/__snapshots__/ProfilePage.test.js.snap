// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="px-5 pt-3 container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col-md-3"
      >
        <h2
          class=" pb-0 mb-0 slide-in"
        >
          User Profile
          <p>
            Update
          </p>
        </h2>
        <form
          class=""
        >
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="name"
            >
              Email Name
            </label>
            <input
              class="form-control"
              id="name"
              placeholder="Enter name"
              type="name"
              value="AdminName"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="email"
            >
              Email Address
            </label>
            <input
              class="form-control"
              id="email"
              placeholder="Enter email"
              type="email"
              value="admin@example.com"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="password"
            >
              Password
            </label>
            <input
              class="form-control"
              id="password"
              placeholder="Enter password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="confirmPassword"
            >
              Confirm Password
            </label>
            <input
              class="form-control"
              id="confirmPassword"
              placeholder="Confirm password"
              type="confirmPassword"
              value=""
            />
          </div>
          <button
            class="btn btn-primary rounded my-2"
            type="submit"
          >
            Update
          </button>
        </form>
      </div>
      <div
        class="col-md-9"
      >
        <h2>
          My Orders
        </h2>
        <div
          class="table-responsive"
        >
          <table
            class="table-sm table table-bordered table-hover"
          >
            <thead>
              <tr
                class="bg-black text-light"
              >
                <th>
                  ID
                </th>
                <th>
                  DATE
                </th>
                <th>
                  TOTAL
                </th>
                <th>
                  PAID
                </th>
                <th>
                  DELIVERED
                </th>
                <th>
                  DETAILS
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  6125d0c2d3ffd29bedee24af
                </td>
                <td>
                  2021-08-25
                </td>
                <td>
                  459.99
                </td>
                <td>
                  2021-08-26
                </td>
                <td>
                  2021-08-26
                </td>
                <td>
                  <a
                    class="btn-sm btn btn-light"
                    href="/orders/6125d0c2d3ffd29bedee24af"
                  >
                    Details
                  </a>
                </td>
              </tr>
              <tr>
                <td>
                  6126a23b644a96d02caece4c
                </td>
                <td>
                  2021-08-25
                </td>
                <td>
                  459.99
                </td>
                <td>
                  <i
                    class="fas fa-times"
                    style="color: red;"
                  />
                </td>
                <td>
                  <i
                    class="fas fa-times"
                    style="color: red;"
                  />
                </td>
                <td>
                  <a
                    class="btn-sm btn btn-light"
                    href="/orders/6126a23b644a96d02caece4c"
                  >
                    Details
                  </a>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
